### ISSUE TITLE
NumericTransformer

### ISSUE THREAD

**Issue by etrain** [amplab/keystone#77]
We're running into a common issue where nodes that should be able to support either DenseVector[Double] or DenseVector[Float] generically are not because breeze supports these two datatypes inconsistently and converting to/from some input type generically is hard. 

@shivaram suggested we create a new "NumericTransformer" API which handles this in a type-safe way and hides some of the complexity under the covers. Things this API might have in it include static checking for input/output types, efficient implicit conversions for the case where conversion is a noop, etc.

More discussion on #74 

cc @tomerk

**Comment by tomerk** [comment]
@etrain how would you guys feel about an interface kind of like this?
```scala
abstract class NumericTransformer[T : BreezeNumeric] extends Transformer[DenseVector[T], DenseVector[T]] {
  def apply(in: DenseVector[T]): DenseVector[T] = {
    apply(in.asDenseMatrix).flatten()
  }

  def apply(in: DenseMatrix[T]): DenseMatrix[T]
  
  def onMatrix: NumericMatrixTransformer[T] = NumericMatrixTransformer(this)
}

abstract class NumericEstimator[T : BreezeNumeric] extends Estimator[DenseVector[T], DenseVector[T]] {
  def fit(data: RDD[DenseVector[T]]): NumericTransformer[T]

  def onMatrix: NumericMatrixEstimator[T] = NumericMatrixEstimator(this)
}

case class NumericMatrixTransformer[T : BreezeNumeric](x: NumericTransformer[T]) extends Transformer[DenseMatrix[T], DenseMatrix[T]] {
  def apply(in: DenseMatrix[T]): DenseMatrix[T] = x.apply(in)
}

case class NumericMatrixEstimator[T : BreezeNumeric](x: NumericEstimator[T]) extends Estimator[DenseMatrix[T], DenseMatrix[T]] {
  def fit(data: RDD[DenseMatrix[T]]): NumericMatrixTransformer[T] = {
    x.fit(data.flatMap(MatrixUtils.matrixToRowArray)).onMatrix
  }
}
```

I'm also starting to wonder if maybe using DenseVector as the canonical features representation instead of a DenseMatrix with only one row is more trouble than it's worth.

**Comment by tomerk** [comment]
Especially because it's easier to translate Matlab code to work on DenseMatrix than DenseVector

**Comment by etrain** [comment]
I think it's a good idea to keep a separation between Matrix and Vector - while a matrix generalizes a vector, it's good for interfaces to reflect the logical intent of the operators, and probably more natural for Transformer developers to be operating on DenseVector than DenseMatrix.

Also - we should shy away from choosing interfaces because it makes translating from Matlab easier. Instead, let's try and focus on having the best possible interfaces so that we don't accidentally assume some of Matlab's deficiencies.

**Comment by tomerk** [comment]
Fair enough. With that in mind how do you feel about the above interfaces?

**Comment by tomerk** [comment]
I guess you would prefer requiring the single vector impl to be implemented and having a default matrix impl instead of the above?

**Comment by etrain** [comment]
Yeah, that could work. In general I thought this issue was about centralized handling f the Double vs. Float problem - is this addressed via `BreezeNumeric` and if so, how does that work?

**Comment by tomerk** [comment]
It is, but there are enough cases where we want to do an "on matrix" / batch operation on what's otherwise a normal NumericTransformer that I think this is a relatively reasonable place to include that functionality (and allow overriding for optimized implementations).

**Comment by etrain** [comment]
well, keep in mind that most of the time we're talking about batching matrix operations together, it's probably going to be about converting an entire partition into a row matrix and allowing for an override of default behavior that operates on that matrix.

**Comment by etrain** [comment]
but yes - i agree this is a fine place for that functionality!

**Comment by tomerk** [comment]
There's that case and there's all the cases in the image pipelines, or e.g. LDA on the frames in the TIMIT pipeline

**Comment by etrain** [comment]
Right, right. I think whatever we figure out for the DenseVector case is also going to work in the DenseMatrix case (w.r.t. the Float/Double thing) and I'm unsure if we want to put those all under the "NumericTransformer" umbrella or not. This actually might be a reasonable place to think about using mixins.

**Comment by tomerk** [comment]
So, I actually looked at Mixins for this previously. To be able to mix in something you have to construct it with the "new" command, so you can't use a case class (or any object) constructor
